File: ios/App/Snappet/Features/Expense/NewExpenseSheet.swift
import SwiftUI
import SwiftData

/// Sheet to add — or edit — an `ExpenseRecord` in a group: title, amount, who paid, and
/// which participants share the cost (default all, equal split). When an existing
/// `record` is passed the form is pre-filled and saving updates it in place. Logs the
/// action to `SnappetCore` on save.
struct NewExpenseSheet: View {
    @Environment(\.modelContext) private var modelContext
    @Environment(\.dismiss) private var dismiss
    @Environment(SnappetCore.self) private var core

    let group: ExpenseGroup
    /// The expense being edited, or `nil` when adding a new one.
    let record: ExpenseRecord?

    @State private var title = ""
    @State private var amount = 0.0
    @State private var payer: String
    /// Names included in the split; defaults to everyone.
    @State private var splitAmong: Set<String>

    init(group: ExpenseGroup, record: ExpenseRecord? = nil) {
        self.group = group
        self.record = record
        _title = State(initialValue: record?.title ?? "")
        _amount = State(initialValue: record?.amount ?? 0.0)
        _payer = State(initialValue: record?.payer ?? group.participants.first ?? "")
        _splitAmong = State(initialValue: Set(record?.participants ?? group.participants))
    }

    private var isEditing: Bool { record != nil }

    private var currencyCode: String { Locale.current.currency?.identifier ?? "USD" }

    var body: some View {
        NavigationStack {
            Form {
                Section("Expense") {
                    TextField("Title (e.g. Dinner)", text: $title)
                        .accessibilityIdentifier("expense.expense.title")
                    TextField("Amount", value: $amount,
                              format: .currency(code: currencyCode))
                        .keyboardType(.decimalPad)
                        .accessibilityIdentifier("expense.expense.amount")
                }

                Section("Paid by") {
                    Picker("Paid by", selection: $payer) {
                        ForEach(group.participants, id: \.self) { person in
                            Text(person).tag(person)
                        }
                    }
                    .pickerStyle(.menu)
                    .accessibilityIdentifier("expense.expense.payer")
                }

                Section {
                    ForEach(group.participants, id: \.self) { person in
                        Button {
                            toggle(person)
                        } label: {
                            HStack {
                                Text(person).foregroundStyle(.primary)
                                Spacer()
                                if splitAmong.contains(person) {
                                    Image(systemName: "checkmark")
                                        .foregroundStyle(.tint)
                                }
                            }
                        }
                        .accessibilityIdentifier("expense.participant.\(person)")
                    }
                } header: {
                    Text("Split equally among")
                } footer: {
                    Text(splitSummary)
                }
            }
            .navigationTitle(isEditing ? "Edit Expense" : "New Expense")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Save") { save() }
                        .disabled(!canSave)
                        .accessibilityIdentifier("expense.expense.save")
                }
            }
        }
    }

    private var canSave: Bool {
        !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && amount > 0
            && !payer.isEmpty
            && !splitAmong.isEmpty
    }

    private var splitSummary: String {
        guard !splitAmong.isEmpty, amount > 0 else {
            return "Select who shares this expense."
        }
        let share = amount / Double(splitAmong.count)
        return "\(share.formatted(.currency(code: currencyCode))) each (\(splitAmong.count) people)."
    }

    private func toggle(_ person: String) {
        if splitAmong.contains(person) {
            splitAmong.remove(person)
        } else {
            splitAmong.insert(person)
        }
    }

    private func save() {
        guard canSave else { return }
        let trimmedTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
        // Preserve the group's participant order for the stored split list.
        let splitList = group.participants.filter { splitAmong.contains($0) }

        if let record {
            // Edit in place — balances recompute from the updated fields.
            record.title = trimmedTitle
            record.amount = amount
            record.payer = payer
            record.participants = splitList
        } else {
            let newRecord = ExpenseRecord(
                groupID: group.id,
                title: trimmedTitle,
                amount: amount,
                payer: payer,
                participants: splitList
            )
            modelContext.insert(newRecord)
        }
        try? modelContext.save()

        let verb = isEditing ? "Edited" : "Added"
        core.log(module: "expense", action: "expense",
                 summary: "\(verb) \(amount.formatted(.currency(code: currencyCode))) expense",
                 metric: amount)
        Haptics.success()
        dismiss()
    }
}

File: android/app/src/main/java/com/snappet/mobile/feature/expense/NewExpenseSheet.kt
package com.snappet.mobile.feature.expense

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.unit.dp

/**
 * Sheet to add — or edit — an [ExpenseRecord] in a group: title, amount, who paid, and which
 * participants share the cost (default all, equal split). When [record] is passed the form is
 * pre-filled and saving updates it in place. The payer defaults to the first participant so the
 * common case needs no picking. Mirrors iOS `NewExpenseSheet`.
 */
@Composable
fun NewExpenseSheet(
    group: ExpenseGroup,
    record: ExpenseRecord?,
    onSave: (title: String, amount: Double, payer: String, splitAmong: List<String>) -> Unit,
) {
    val participantsList = group.participants
    var title by remember { mutableStateOf(record?.title ?: "") }
    var amountText by remember { mutableStateOf(record?.amount?.let { formatAmount(it) } ?: "") }
    var payer by remember { mutableStateOf(record?.payer ?: participantsList.firstOrNull() ?: "") }
    val splitAmong = remember {
        mutableStateListOf<String>().apply { addAll(record?.participants ?: participantsList) }
    }

    val amount = amountText.toDoubleOrNull() ?: 0.0
    val canSave = title.trim().isNotEmpty() && amount > 0 && payer.isNotEmpty() && splitAmong.isNotEmpty()

    Column(
        Modifier.fillMaxWidth().verticalScroll(rememberScrollState()).padding(24.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        Text(
            if (record != null) "Edit Expense" else "New Expense",
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.Bold,
        )
        OutlinedTextField(
            value = title,
            onValueChange = { title = it },
            label = { Text("Title (e.g. Dinner)") },
            singleLine = true,
            modifier = Modifier.fillMaxWidth().testTag("expense.expense.title"),
        )
        OutlinedTextField(
            value = amountText,
            onValueChange = { amountText = it },
            label = { Text("Amount") },
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
            modifier = Modifier.fillMaxWidth().testTag("expense.expense.amount"),
        )

        Text("Paid by", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
        ParticipantPicker(
            participants = participantsList,
            selected = payer,
            tag = "expense.expense.payer",
            onSelect = { payer = it },
        )

        Text("Split equally among", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
        participantsList.forEach { person ->
            val checked = splitAmong.contains(person)
            Row(
                Modifier
                    .fillMaxWidth()
                    .clickable {
                        if (checked) splitAmong.remove(person) else splitAmong.add(person)
                    }
                    .padding(vertical = 8.dp)
                    .testTag("expense.participant.$person"),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(person, modifier = Modifier.weight(1f))
                if (checked) {
                    Icon(Icons.Filled.Check, contentDescription = "Included", tint = MaterialTheme.colorScheme.primary)
                }
            }
        }

        Button(
            onClick = {
                if (canSave) {
                    // Preserve the group's participant order for the stored split list.
                    val splitList = participantsList.filter { splitAmong.contains(it) }
                    onSave(title.trim(), amount, payer, splitList)
                }
            },
            enabled = canSave,
            modifier = Modifier.fillMaxWidth().testTag("expense.expense.save"),
        ) {
            Text("Save")
        }
    }
}
